File: modern-jogl-examples/src/tut15/manyImages/ManyImages.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package tut15.manyImages;

import com.jogamp.newt.awt.NewtCanvasAWT;
import com.jogamp.newt.event.KeyEvent;
import com.jogamp.newt.event.KeyListener;
import com.jogamp.newt.opengl.GLWindow;
import com.jogamp.opengl.util.FPSAnimator;
import com.jogamp.opengl.util.GLBuffers;
import com.jogamp.opengl.util.texture.TextureData;
import com.jogamp.opengl.util.texture.TextureIO;
import com.jogamp.opengl.util.texture.spi.DDSImage;
import com.jogamp.opengl.util.texture.spi.DDSImage.ImageInfo;
import glutil.MatrixStack;
import glutil.Timer;
import java.awt.Frame;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.nio.Buffer;
import java.nio.ByteBuffer;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.media.opengl.GL3;
import javax.media.opengl.GLAutoDrawable;
import javax.media.opengl.GLCapabilities;
import javax.media.opengl.GLEventListener;
import javax.media.opengl.GLProfile;
import jglm.Jglm;
import jglm.Mat4;
import jglm.Vec3;
import mesh.Mesh;
import tut15.manyImages.program.ProgramData;

/**
 *
 * @author gbarbieri
 */
public class ManyImages implements GLEventListener, KeyListener {

    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {

        final ManyImages manyImages = new ManyImages();
        manyImages.initGL();

        Frame frame = new Frame("Tutorial 15 - Many Images");

        frame.add(manyImages.newtCanvasAWT);

        frame.setSize(manyImages.glWindow.getWidth(), manyImages.glWindow.getHeight());

        final FPSAnimator fPSAnimator = new FPSAnimator(manyImages.glWindow, 30);

        frame.addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosing(WindowEvent windowEvent) {
                fPSAnimator.stop();
                manyImages.glWindow.destroy();
                System.exit(0);
            }
        });

        fPSAnimator.start();
        frame.setVisible(true);
    }

    private GLWindow glWindow;
    private NewtCanvasAWT newtCanvasAWT;
    private ProgramData program;
    private Mesh plane;
    private Mesh corridor;
    private int[] uniformBlockBuffers = new int[UniformBlockBinding.size.ordinal()];
    private int[] textures = new int[Textures.size.ordinal()];
    private int[] samplers;
    private Timer camTimer = new Timer(Timer.Type.Loop, 5f);
    private boolean useMipmapTexture = false;
    private int currSampler = 0;
    private boolean drawCorridor = false;

    public ManyImages() {

    }

    public void initGL() {

        GLProfile gLProfile = GLProfile.get(GLProfile.GL3);

        GLCapabilities gLCapabilities = new GLCapabilities(gLProfile);

        glWindow = GLWindow.create(gLCapabilities);

        glWindow.setSize(1024, 768);

        glWindow.addGLEventListener(this);
        glWindow.addKeyListener(this);
//        glWindow.addMouseListener(this);
        /*
         *  We combine NEWT GLWindow inside existing AWT application (the main JFrame) 
         *  by encapsulating the glWindow inside a NewtCanvasAWT canvas.
         */
        newtCanvasAWT = new NewtCanvasAWT(glWindow);
    }

    @Override
    public void init(GLAutoDrawable glad) {

        GL3 gl3 = glad.getGL().getGL3();

        initializeProgram(gl3);

        String dataPath = "/tut15/manyImages/data/";

        plane = new Mesh(dataPath + "BigPlane.xml", gl3);
        corridor = new Mesh(dataPath + "Corridor.xml", gl3);

        gl3.glEnable(GL3.GL_CULL_FACE);
        gl3.glCullFace(GL3.GL_BACK);
        gl3.glFrontFace(GL3.GL_CW);

        gl3.glEnable(GL3.GL_DEPTH_TEST);
        gl3.glDepthMask(true);
        gl3.glDepthFunc(GL3.GL_LEQUAL);
        gl3.glDepthRangef(0f, 1f);
        gl3.glEnable(GL3.GL_DEPTH_CLAMP);

        //Setup our Uniform Buffers
        gl3.glGenBuffers(1, uniformBlockBuffers, UniformBlockBinding.projection.ordinal());
        gl3.glBindBuffer(GL3.GL_UNIFORM_BUFFER, uniformBlockBuffers[UniformBlockBinding.projection.ordinal()]);
        {
            gl3.glBufferData(GL3.GL_UNIFORM_BUFFER, 16 * 4, null, GL3.GL_DYNAMIC_DRAW);

            gl3.glBindBufferRange(GL3.GL_UNIFORM_BUFFER, UniformBlockBinding.projection.ordinal(),
                    uniformBlockBuffers[UniformBlockBinding.projection.ordinal()], 0, 16 * 4);
        }
        gl3.glBindBuffer(GL3.GL_UNIFORM_BUFFER, 0);

        loadCheckerTexture(gl3, dataPath);

        loadMipmapTexture(gl3);

        createSamplers(gl3);
    }

    private void initializeProgram(GL3 gl3) {

        String shadersPath = "/tut15/manyImages/shaders/";

        program = new ProgramData(gl3, shadersPath, "PT_VS.glsl", "Tex_FS.glsl");
    }

    private void loadCheckerTexture(GL3 gl3, String dataPath) {

        URL url = getClass().getResource(dataPath + "checker.dds");
        File file = new File(url.getPath());

        DDSImage ddsImage = null;

        try {
            ddsImage = DDSImage.read(file);
        } catch (IOException ex) {
            Logger.getLogger(ManyImages.class.getName()).log(Level.SEVERE, null, ex);
        }

        gl3.glGenTextures(1, textures, Textures.checker.ordinal());
        gl3.glBindTexture(GL3.GL_TEXTURE_2D, textures[Textures.checker.ordinal()]);
        {
            for (int mipmapLevel = 0; mipmapLevel < ddsImage.getNumMipMaps(); mipmapLevel++) {

                ImageInfo mipmap = ddsImage.getMipMap(mipmapLevel);

                gl3.glTexImage2D(GL3.GL_TEXTURE_2D, mipmapLevel, GL3.GL_RGB8, mipmap.getWidth(), mipmap.getHeight(),
                        0, GL3.GL_BGRA, GL3.GL_UNSIGNED_INT_8_8_8_8_REV, mipmap.getData());
            }

            gl3.glTexParameteri(GL3.GL_TEXTURE_2D, GL3.GL_TEXTURE_BASE_LEVEL, 0);
            gl3.glTexParameteri(GL3.GL_TEXTURE_2D, GL3.GL_TEXTURE_MAX_LEVEL, ddsImage.getNumMipMaps() - 1);
        }
        gl3.glBindTexture(GL3.GL_TEXTURE_2D, 0);
    }

    private void loadMipmapTexture(GL3 gl3) {

        byte[] mipmapColors = new byte[]{
            (byte) 0xFF, (byte) 0xFF, (byte) 0x00,
            (byte) 0xFF, (byte) 0x00, (byte) 0xFF,
            (byte) 0x00, (byte) 0xFF, (byte) 0xFF,
            (byte) 0xFF, (byte) 0x00, (byte) 0x00,
            (byte) 0x00, (byte) 0xFF, (byte) 0x00,
            (byte) 0x00, (byte) 0x00, (byte) 0xFF,
            (byte) 0x00, (byte) 0x00, (byte) 0x00,
            (byte) 0xFF, (byte) 0xFF, (byte) 0xFF};

        gl3.glGenTextures(1, textures, Textures.mipmapTest.ordinal());

        gl3.glBindTexture(GL3.GL_TEXTURE_2D, textures[Textures.mipmapTest.ordinal()]);
        {
            int[] oldAlign = new int[]{0};

            gl3.glGetIntegerv(GL3.GL_UNPACK_ALIGNMENT, oldAlign, 0);
            gl3.glPixelStorei(GL3.GL_UNPACK_ALIGNMENT, 1);
            {
                for (int mipmapLevel = 0; mipmapLevel < 8; mipmapLevel++) {

                    int width = 128 >> mipmapLevel;
                    int height = 128 >> mipmapLevel;

                    byte[] buffer = fillWithColors(mipmapColors[mipmapLevel * 3 + 0],
                            mipmapColors[mipmapLevel * 3 + 1], mipmapColors[mipmapLevel * 3 + 2], width, height);

                    ByteBuffer byteBuffer = GLBuffers.newDirectByteBuffer(buffer);

                    gl3.glTexImage2D(GL3.GL_TEXTURE_2D, mipmapLevel, GL3.GL_RGB8, width, height,
                            0, GL3.GL_RGB, GL3.GL_UNSIGNED_BYTE, byteBuffer);
                }
            }
            gl3.glPixelStorei(GL3.GL_UNPACK_ALIGNMENT, oldAlign[0]);

            gl3.glTexParameteri(GL3.GL_TEXTURE_2D, GL3.GL_TEXTURE_BASE_LEVEL, 0);
            gl3.glTexParameteri(GL3.GL_TEXTURE_2D, GL3.GL_TEXTURE_MAX_LEVEL, 7);
        }
        gl3.glBindTexture(GL3.GL_TEXTURE_2D, 0);
    }

    private void createSamplers(GL3 gl3) {

        samplers = new int[6];

        gl3.glGenSamplers(6, samplers, 0);

        for (int samplerIx = 0; samplerIx < 6; samplerIx++) {

            gl3.glSamplerParameteri(samplers[samplerIx], GL3.GL_TEXTURE_WRAP_S, GL3.GL_REPEAT);
            gl3.glSamplerParameteri(samplers[samplerIx], GL3.GL_TEXTURE_WRAP_T, GL3.GL_REPEAT);
        }
        //Nearest
        gl3.glSamplerParameteri(samplers[0], GL3.GL_TEXTURE_MAG_FILTER, GL3.GL_NEAREST);
        gl3.glSamplerParameteri(samplers[0], GL3.GL_TEXTURE_MIN_FILTER, GL3.GL_NEAREST);

        //Linear
        gl3.glSamplerParameteri(samplers[1], GL3.GL_TEXTURE_MAG_FILTER, GL3.GL_LINEAR);
        gl3.glSamplerParameteri(samplers[1], GL3.GL_TEXTURE_MIN_FILTER, GL3.GL_LINEAR);

        //Linear mipmap Nearest
        gl3.glSamplerParameteri(samplers[2], GL3.GL_TEXTURE_MAG_FILTER, GL3.GL_LINEAR);
        gl3.glSamplerParameteri(samplers[2], GL3.GL_TEXTURE_MIN_FILTER, GL3.GL_LINEAR_MIPMAP_NEAREST);

        //Linear mipmap linear
        gl3.glSamplerParameteri(samplers[3], GL3.GL_TEXTURE_MAG_FILTER, GL3.GL_LINEAR);
        gl3.glSamplerParameteri(samplers[3], GL3.GL_TEXTURE_MIN_FILTER, GL3.GL_LINEAR_MIPMAP_LINEAR);

        //Low anisotropic
        gl3.glSamplerParameteri(samplers[4], GL3.GL_TEXTURE_MAG_FILTER, GL3.GL_LINEAR);
        gl3.glSamplerParameteri(samplers[4], GL3.GL_TEXTURE_MIN_FILTER, GL3.GL_LINEAR_MIPMAP_LINEAR);
        gl3.glSamplerParameterf(samplers[4], GL3.GL_TEXTURE_MAX_ANISOTROPY_EXT, 4.0f);

        //Max anisotropic
        float[] maxAniso = new float[]{0};

        gl3.glGetFloatv(GL3.GL_MAX_TEXTURE_MAX_ANISOTROPY_EXT, maxAniso, 0);

        System.out.println("Maximum anisotropy: " + maxAniso[0]);

        gl3.glSamplerParameteri(samplers[5], GL3.GL_TEXTURE_MAG_FILTER, GL3.GL_LINEAR);
        gl3.glSamplerParameteri(samplers[5], GL3.GL_TEXTURE_MIN_FILTER, GL3.GL_LINEAR_MIPMAP_LINEAR);
        gl3.glSamplerParameteri(samplers[5], GL3.GL_TEXTURE_MAX_ANISOTROPY_EXT, (int) maxAniso[0]);
    }

    private byte[] fillWithColors(byte red, byte green, byte blue, int width, int height) {

        int numTexels = width * height;
        byte[] buffer = new byte[numTexels * 3];

        for (int i = 0; i < numTexels; i++) {

            buffer[i * 3 + 0] = red;
            buffer[i * 3 + 1] = green;
            buffer[i * 3 + 2] = blue;
        }
        return buffer;
    }

    @Override
    public void dispose(GLAutoDrawable glad) {

    }

    @Override
    public void display(GLAutoDrawable glad) {

        GL3 gl3 = glad.getGL().getGL3();

        gl3.glClearColor(.75f, .75f, 1f, 1f);
        gl3.glClearDepth(1f);
        gl3.glClear(GL3.GL_COLOR_BUFFER_BIT | GL3.GL_DEPTH_BUFFER_BIT);

        if (plane != null && corridor != null) {

            camTimer.update();

            float cyclicAngle = camTimer.getAlpha() * 6.28f;
            float hOffset = (float) (Math.cos(cyclicAngle) * .25f);
            float vOffset = (float) (Math.sin(cyclicAngle) * .25f);

            MatrixStack modelMatrix = new MatrixStack();

            Mat4 worldToCamMat = Mat4.CalcLookAtMatrix(new Vec3(hOffset, 1f, -64f),
                    new Vec3(hOffset, -5f + vOffset, -44f), new Vec3(0f, 1f, 0f));

            modelMatrix.setTop(worldToCamMat);

            modelMatrix.push();
            {
                program.bind(gl3);
                {
                    gl3.glUniformMatrix4fv(program.getModelToCameraMatrixUL(), 1, false,
                            modelMatrix.top().toFloatArray(), 0);

                    gl3.glActiveTexture(GL3.GL_TEXTURE0);
                    gl3.glBindTexture(GL3.GL_TEXTURE_2D,
                            textures[useMipmapTexture ? Textures.mipmapTest.ordinal() : Textures.checker.ordinal()]);
                    {
                        gl3.glBindSampler(TexUnit.color.ordinal(), samplers[currSampler]);
                        {
                            if (drawCorridor) {

                                corridor.render(gl3, "tex");

                            } else {

                                plane.render(gl3, "tex");
                            }
                        }
                        gl3.glBindSampler(TexUnit.color.ordinal(), 0);
                    }
                    gl3.glBindTexture(GL3.GL_TEXTURE_2D, 0);
                }
                program.unbind(gl3);
            }
            modelMatrix.pop();
        }
    }

    @Override
    public void reshape(GLAutoDrawable glad, int x, int y, int w, int h) {

        GL3 gl3 = glad.getGL().getGL3();

        Mat4 persMatrix = Jglm.perspective(90f, w / (float) h, 1f, 1000f);

        gl3.glBindBuffer(GL3.GL_UNIFORM_BUFFER, uniformBlockBuffers[UniformBlockBinding.projection.ordinal()]);
        {
            Buffer buffer = GLBuffers.newDirectFloatBuffer(persMatrix.toFloatArray());

            gl3.glBufferSubData(GL3.GL_UNIFORM_BUFFER, 0, 16 * 4, buffer);
        }
        gl3.glBindBuffer(GL3.GL_UNIFORM_BUFFER, 0);

        gl3.glViewport(x, y, w, h);
    }

    @Override
    public void keyPressed(KeyEvent ke) {

        switch (ke.getKeyCode()) {

            case KeyEvent.VK_SPACE:
                useMipmapTexture = !useMipmapTexture;
                break;

            case KeyEvent.VK_Y:
                drawCorridor = !drawCorridor;
                break;

            case KeyEvent.VK_P:
                camTimer.togglePause();
                break;

            case KeyEvent.VK_1:
                currSampler = 0;
                break;

            case KeyEvent.VK_2:
                currSampler = 1;
                break;

            case KeyEvent.VK_3:
                currSampler = 2;
                break;

            case KeyEvent.VK_4:
                currSampler = 5;
                break;

            case KeyEvent.VK_5:
                currSampler = 4;
                break;

            case KeyEvent.VK_6:
                currSampler = 5;
                break;
        }
    }

    @Override
    public void keyReleased(KeyEvent ke) {

    }

    public enum UniformBlockBinding {

        projection,
        size
    }

    public enum TexUnit {

        color,
        size
    }

    public enum Textures {

        checker,
        mipmapTest,
        size
    }
}

File: src/main/kotlin/glNext/tut15/manyImages.kt
package glNext.tut15

import com.jogamp.newt.event.KeyEvent
import com.jogamp.opengl.GL2ES3.*
import com.jogamp.opengl.GL2GL3.GL_UNSIGNED_INT_8_8_8_8_REV
import com.jogamp.opengl.GL3
import com.jogamp.opengl.util.texture.spi.DDSImage
import glNext.*
import glm.*
import main.framework.Framework
import main.framework.Semantic
import main.framework.component.Mesh
import uno.buffer.*
import uno.glm.MatrixStack
import uno.glsl.programOf
import uno.time.Timer
import java.io.File
import java.nio.ByteBuffer
import glm.mat.Mat4
import glm.vec._3.Vec3

/**
 * Created by GBarbieri on 31.03.2017.
 */

fun main(args: Array<String>) {
    ManyImages_Next().setup("Tutorial 15 - Many Images")
}

class ManyImages_Next : Framework() {

    lateinit var program: ProgramData

    lateinit var plane: Mesh
    lateinit var corridor: Mesh

    val projBufferName = intBufferBig(1)

    object Texture {
        val Checker = 0
        val MipmapTest = 1
        val MAX = 2
    }

    val textureName = intBufferBig(Texture.MAX)
    val samplerName = intBufferBig(Sampler.MAX)

    val camTimer = Timer(Timer.Type.Loop, 5f)
    var useMipmapTexture = false
    var currSampler = 0
    var drawCorridor = false

    override fun init(gl: GL3) = with(gl) {

        initializeProgram(gl)

        plane = Mesh(gl, javaClass, "tut15/BigPlane.xml")
        corridor = Mesh(gl, javaClass, "tut15/Corridor.xml")

        cullFace {
            enable()
            cullFace = back
            frontFace = cw
        }

        val depthZNear = 0f
        val depthZFar = 1f

        depth {
            test = true
            mask = true
            func = lEqual
            rangef = depthZNear..depthZFar
            clamp = true
        }

        //Setup our Uniform Buffers
        initUniformBuffer(projBufferName) {

            data(Mat4.SIZE, GL_DYNAMIC_DRAW)

            range(Semantic.Uniform.PROJECTION, 0, Mat4.SIZE)
        }

        // Generate all the texture names
        glGenTextures(textureName)

        loadCheckerTexture(gl)
        loadMipmapTexture(gl)
        createSamplers(gl)
    }

    fun initializeProgram(gl: GL3) {
        program = ProgramData(gl, "pt.vert", "tex.frag")
    }

    fun loadCheckerTexture(gl: GL3) = with(gl) {

        val file = File(javaClass.getResource("/tut15/checker.dds").toURI())

        val ddsImage = DDSImage.read(file)

        withTexture2d(textureName[Texture.Checker]) {

            repeat(ddsImage.numMipMaps) { mipmapLevel ->

                val mipmap = ddsImage.getMipMap(mipmapLevel)

                image(mipmapLevel, GL_RGB8, mipmap.width, mipmap.height, GL_BGRA, GL_UNSIGNED_INT_8_8_8_8_REV, mipmap.data)
            }
            levels = 0 until ddsImage.numMipMaps
        }
    }

    fun loadMipmapTexture(gl: GL3) = with(gl) {

        withTexture2d(textureName[Texture.MipmapTest]) {

            val oldAlign = glGetInteger(GL_UNPACK_ALIGNMENT)
            glPixelStorei(GL_UNPACK_ALIGNMENT, 1)

            for (mipmapLevel in 0..7) {

                val width = 128 shr mipmapLevel
                val height = 128 shr mipmapLevel

                val currColor = mipmapColors[mipmapLevel]
                val buffer = fillWithColors(currColor, width, height)

                image(mipmapLevel, GL_RGB8, width, height, GL_RGB, GL_UNSIGNED_BYTE, buffer)

                buffer.destroy()
            }
            glPixelStorei(GL_UNPACK_ALIGNMENT, oldAlign)
            levels = 0..7
        }
    }

    val mipmapColors = arrayOf(
            byteArrayOf(0xFF.b, 0xFF.b, 0x00.b),
            byteArrayOf(0xFF.b, 0x00.b, 0xFF.b),
            byteArrayOf(0x00.b, 0xFF.b, 0xFF.b),
            byteArrayOf(0xFF.b, 0x00.b, 0x00.b),
            byteArrayOf(0x00.b, 0xFF.b, 0x00.b),
            byteArrayOf(0x00.b, 0x00.b, 0xFF.b),
            byteArrayOf(0x00.b, 0x00.b, 0x00.b),
            byteArrayOf(0xFF.b, 0xFF.b, 0xFF.b))

    fun fillWithColors(color: ByteArray, width: Int, height: Int): ByteBuffer {

        val numTexels = width * height
        val buffer = byteBufferBig(numTexels * 3)

        val (red, green, blue) = color

        while (buffer.hasRemaining())
            buffer
                    .put(red)
                    .put(green)
                    .put(blue)

        buffer.position(0)
        return buffer
    }

    fun createSamplers(gl: GL3) = with(gl) {

        initSamplers(samplerName) {

            for (i in 0 until Sampler.MAX)
                at(i) {
                    wrapS = repeat
                    wrapT = repeat
                }
            at(Sampler.Nearest) {
                magFilter = nearest
                minFilter = nearest
            }
            at(Sampler.Linear) {
                magFilter = linear
                minFilter = linear
            }
            at(Sampler.Linear_MipMap_Nearest) {
                magFilter = linear_mmNearest
                minFilter = linear_mmNearest
            }
            at(Sampler.Linear_MipMap_Linear) {
                magFilter = linear_mmLinear
                minFilter = linear_mmLinear
            }
            at(Sampler.LowAnisotropy) {
                magFilter = linear
                minFilter = linear_mmLinear
                maxAnisotropy = 4.0f
            }

            val maxAniso = caps.limits.MAX_TEXTURE_MAX_ANISOTROPY_EXT // TODO float?
            println("Maximum anisotropy: " + maxAniso)

            at(Sampler.MaxAnisotropy) {
                magFilter = linear
                minFilter = linear_mmLinear
                maxAnisotropy = maxAniso.f
            }
        }
    }

    override fun display(gl: GL3) = with(gl) {

        clear {
            color(0.75f, 0.75f, 1.0f, 1.0f)
            depth()
        }

        camTimer.update()

        val cyclicAngle = camTimer.getAlpha() * 6.28f
        val hOffset = glm.cos(cyclicAngle) * .25f
        val vOffset = glm.sin(cyclicAngle) * .25f

        val modelMatrix = MatrixStack()

        val worldToCamMat = glm.lookAt(
                Vec3(hOffset, 1f, -64f),
                Vec3(hOffset, -5f + vOffset, -44f),
                Vec3(0f, 1f, 0f))

        modelMatrix.applyMatrix(worldToCamMat) run {

            usingProgram(program.theProgram) {

                program.modelToCameraMatrixUL.mat4 = top()

                val texture = textureName[if (useMipmapTexture) Texture.MipmapTest else Texture.Checker]
                withTexture2d(Semantic.Sampler.DIFFUSE, texture, samplerName[currSampler]) {

                    if (drawCorridor)
                        corridor.render(gl, "tex")
                    else
                        plane.render(gl, "tex")
                }
            }
        }
    }

    override fun reshape(gl: GL3, w: Int, h: Int) = with(gl) {

        val persMatrix = MatrixStack()
        persMatrix.perspective(90f, w / h.f, 1f, 1000f)

        withUniformBuffer(projBufferName) { subData(persMatrix.top()) }

        glViewport(w, h)
    }

    override fun end(gl: GL3) = with(gl) {

        plane.dispose(gl)
        corridor.dispose(gl)

        glDeleteProgram(program.theProgram)

        glDeleteBuffer(projBufferName)
        glDeleteTextures(textureName)
        glDeleteSamplers(samplerName)

        destroyBuffers(projBufferName, textureName, samplerName)
    }

    override fun keyPressed(ke: KeyEvent) {

        when (ke.keyCode) {

            KeyEvent.VK_ESCAPE -> quit()

            KeyEvent.VK_SPACE -> useMipmapTexture = !useMipmapTexture

            KeyEvent.VK_Y -> drawCorridor = !drawCorridor

            KeyEvent.VK_P -> camTimer.togglePause()
        }

        if (ke.keyCode in KeyEvent.VK_1..KeyEvent.VK_9) {
            val number = ke.keyCode - KeyEvent.VK_1
            if (number < Sampler.MAX) {
                println("Sampler: " + samplerNames[number])
                currSampler = number
            }
        }
    }

    object Sampler {
        val Nearest = 0
        val Linear = 1
        val Linear_MipMap_Nearest = 2
        val Linear_MipMap_Linear = 3
        val LowAnisotropy = 4
        val MaxAnisotropy = 5
        val MAX = 6
    }

    val samplerNames = arrayOf("Nearest", "Linear", "Linear with nearest mipmaps", "Linear with linear mipmaps", "Low anisotropic", "Max anisotropic")

    class ProgramData(gl: GL3, vertex: String, fragment: String) {

        val theProgram = programOf(gl, javaClass, "tut15", vertex, fragment)
        val modelToCameraMatrixUL = gl.glGetUniformLocation(theProgram, "modelToCameraMatrix")

        init {
            with(gl) {
                glUniformBlockBinding(
                        theProgram,
                        glGetUniformBlockIndex(theProgram, "Projection"),
                        Semantic.Uniform.PROJECTION)

                glUseProgram(theProgram)
                glUniform1i(
                        glGetUniformLocation(theProgram, "colorTexture"),
                        Semantic.Sampler.DIFFUSE)
                glUseProgram()
            }
        }
    }
}